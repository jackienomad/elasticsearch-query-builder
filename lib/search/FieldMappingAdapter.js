export class FieldMappingAdapter {
  /*------------------------------CONSTRUCTOR------------------------------*/
  constructor(fieldMapping) {
    if (!fieldMapping) throw new Error('fieldMapping is required')
    this.fieldMapping = fieldMapping
  }

  destroy() {
    this.fieldMapping = null
  }

  /*------------------------------METHODS------------------------------*/
  getMappedField(selectedField, isUntouched = false) {
    if (!this.fieldMapping[selectedField]) return ''

    if (isUntouched) {
      return (
        this.fieldMapping[selectedField].not_analyzed_field ||
        this.fieldMapping[selectedField].field ||
        ''
      )
    } else {
      return (
        this.fieldMapping[selectedField].field ||
        this.fieldMapping[selectedField].not_analyzed_field ||
        ''
      )
    }
  }

  getMiscField(miscField) {
    if (!this.fieldMapping[miscField]) return ''
    const correspondingField = this.fieldMapping[miscField].field

    if (!this.fieldMapping[correspondingField]) return correspondingField
    return this.fieldMapping[correspondingField].field || ''
  }
}
export default FieldMappingAdapter
