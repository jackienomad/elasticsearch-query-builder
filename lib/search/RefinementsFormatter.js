export class RefinementsFormatter {
  static format(response, refinementGroups) {
    return Object.keys(response.aggregations)
      .map(field => {
        const refinementGroup = {
          field,
          category: refinementGroups[field]
        }

        if (response.aggregations[field].buckets) {
          refinementGroup.totalRefinements =
            response.aggregations[field].buckets.length
          refinementGroup.refinements = response.aggregations[
            field
          ].buckets.map(({ key: name, doc_count: count }) => ({
            name,
            count
          }))
        } else {
          // miscellaneous
          refinementGroup.totalCount =
            response.aggregations[field].doc_count || 0
        }

        return refinementGroup
      })
      .sort(({ field: field1 }, { field: field2 }) => {
        const ucField1 = field1.toUpperCase()
        const ucField2 = field2.toUpperCase()
        if (ucField1 < ucField2) {
          return -1
        } else if (ucField2 < ucField1) {
          return 1
        }

        return 0
      })
  }
}

export default RefinementsFormatter
