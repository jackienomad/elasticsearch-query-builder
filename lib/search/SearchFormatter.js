export class SearchFormatter {
  static formatResponse(esResponse) {
    return esResponse.hits.hits.map(({ _id, _source }) => ({
      ArrivalDate: _source['Arrival Date'] || '',
      BillOfLading: _source['Bill of Lading Number'] || '',
      BillTypeCode: _source['Bill Type'] || '',
      CarrierAddress: _source['Carrier']['Address'] || '',
      CarrierCity: (_source['Carrier'] && _source['Carrier']['City']) || '',
      CarrierCode: (_source['Carrier'] && _source['Carrier']['Code']) || '',
      CarrierName: (_source['Carrier'] && _source['Carrier']['Name']) || '',
      CarrierState: (_source['Carrier'] && _source['Carrier']['State']) || '',
      CarrierZip: (_source['Carrier'] && _source['Carrier']['Zip Code']) || '',
      Consignee: (_source['Consignee'] && _source['Consignee']['Name']) || '',
      ConsigneeAddress:
        (_source['Consignee'] && _source['Consignee']['Address']) || '',
      ConsigneeAddressBreakdown:
        _source['Consignee'] &&
        _source['Consignee']['Cleaned Address Breakdown'],
      ConsigneeCity:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['PlaceName']) ||
        '',
      ConsigneeId:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['_id']) ||
        '',
      ConsigneeState:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['StateName']) ||
        '',
      ConsigneeStreet:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['StreetName']) ||
        '',
      ConsigneeStreetNumber:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['AddressNumber']) ||
        '',
      ConsigneeZipCode:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['ZipCode']) ||
        '',
      ContainerHeight:
        (_source['Containers']['0'] && _source['Containers']['0']['Height']) ||
        '',
      ContainerLength:
        (_source['Containers']['0'] && _source['Containers']['0']['Length']) ||
        '',
      ContainerNumber:
        (_source['Containers']['0'] &&
          _source['Containers']['0']['Container Number']) ||
        '',
      ContainerType:
        (_source['Containers']['0'] && _source['Containers']['0']['Type']) ||
        '',
      ContainerWidth:
        (_source['Containers']['0'] && _source['Containers']['0']['Width']) ||
        '',
      Containers: _source['Containers'] || '',
      CountryOfOrigin: _source['Country of Origin'] || '',
      DestinationPort: _source['Port of Destination'] || '',
      EquipmentDescriptionCode:
        (_source['Containers']['0'] &&
          _source['Containers']['0']['Equipment Description']) ||
        '',
      ForeignPort: _source['Foreign Port of Lading'] || '',
      GrossWeightKgs: _source['Weight'] || '',
      GrossWeightLbs: _source['Weight'] || '',
      HSCode: _source['HS Code'] || '',
      InBondEntryType: _source['In-Bond Entry Type'] || '',
      LoadStatus:
        _source['Containers']['0'] && _source['Containers']['0']['Load Status'],
      ManifestNumber: _source['Manifest Number'] || '',
      MarksAndNumbers:
        (_source['Containers']['0'] && _source['Containers']['0']['Marks']) ||
        '',
      MasterBillOfLading: _source['Master Bill of Lading Number'] || '',
      Measurement: _source['Measurement'] || '',
      MeasurementUnit: _source['Measurement Unit'] || '',
      NotifyParty:
        (_source['Notify Party'] && _source['Notify Party']['Name']) || '',
      NotifyPartyAddress:
        (_source['Notify Party'] && _source['Notify Party']['Address']) || '',
      NumberOfContainers: _source['Number of Containers'] || '',
      PlaceOfReceipt: _source['Place of Receipt'] || '',
      ProductDescription:
        (_source['Containers']['0']['Cargoes'] &&
          _source['Containers']['0']['Cargoes']['Description']) ||
        '',
      ProductKeywords:
        (_source['Containers']['0']['Cargoes'] &&
          _source['Containers']['0']['Cargoes']['Keywords']) ||
        '',
      Quantity: _source['Manifest Quantity'] || '',
      QuantityUnit: _source['Manifest Unit'] || '',
      SealNumbers:
        _source['Containers']['0'] &&
        _source['Containers']['0']['Seal Numbers'],
      ShipmentId: _id || '',
      Shipper: (_source['Shipper'] && _source['Shipper']['Name']) || '',
      ShipperAddress:
        (_source['Shipper'] && _source['Shipper']['Address']) || '',
      ShipperId: (_source['Shipper'] && _source['Shipper']['_id']) || '',
      TypeOfService:
        (_source['Containers'] && _source['Containers']['Type of Service']) ||
        '',
      VesselName: _source['Vessel Name'] || '',
      VoyageNumber: _source['Voyage Number'] || ''
    }))
  }
}

export default SearchFormatter
