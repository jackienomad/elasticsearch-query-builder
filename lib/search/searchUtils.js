export const refinementIndexes = ['keyword', 'field']
export const keywordIndexes = ['keyword', 'operator', 'type', 'field']

export const Operators = {
  AND: 'and',
  OR: 'or',
  NOT: 'not'
}
export const SearchTypes = {
  EXACT: 'exact',
  FUZZY: 'fuzzy',
  REGULAR: 'regular'
}
export const MiscTypes = {
  BLANK_CONSIGNEES_EXCLUDED: 'BlankConsigneesExcluded',
  BLANK_SHIPPERS_EXCLUDED: 'BlankShippersExcluded',
  MASTER_SHIPMENTS_EXCLUDED: 'MasterShipmentsExcluded',
  HOUSE_SHIPMENTS_ONLY: 'HouseShipmentsOnly',
  MASTER_SHIPMENTS_ONLY: 'MasterShipmentsOnly'
}

export const percentByType = {
  [SearchTypes.REGULAR]: '100%',
  [SearchTypes.FUZZY]: '75%'
}

export function getMinimumByType(type) {
  return percentByType[type]
}

export function transformArray(transformee, indexes) {
  if (transformee === undefined) {
    return []
  }
  return transformee.map(val => {
    let values = val.split(',')

    if (values.length > indexes.length) {
      const excess = values.splice(
        values.length - indexes.length + 1,
        values.length
      )
      values = [values.join(','), ...excess]
    }

    const obj = {}
    indexes.forEach((prop, index) => {
      obj[prop] = values[index]
    })
    return obj
  })
}
