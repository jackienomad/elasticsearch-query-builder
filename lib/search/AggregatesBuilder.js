import { FieldMappingAdapter } from './FieldMappingAdapter'

export class AggregatesBuilder {
  static build(fields, fieldMapping) {
    const fieldAdapter =
      fieldMapping instanceof FieldMappingAdapter
        ? fieldMapping
        : new FieldMappingAdapter(fieldMapping)

    const aggregations = {}
    fields.forEach(field => {
      aggregations[field] = {
        terms: {
          field: fieldAdapter.getMappedField(field, true),
          size: 100
        }
      }
    })

    return aggregations
  }
}

export default AggregatesBuilder
