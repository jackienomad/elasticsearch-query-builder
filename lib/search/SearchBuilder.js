import {
  getMinimumByType,
  Operators,
  SearchTypes,
  MiscTypes
} from './searchUtils'
import { FieldMappingAdapter } from './FieldMappingAdapter'

export class SearchBuilder {
  /*------------------------------CONSTRUCTOR------------------------------*/
  constructor(index, type, fieldMapping) {
    this.index = index
    this.type = type
    this.fieldMapping = fieldMapping
    this.fieldMappingAdapter = new FieldMappingAdapter(fieldMapping)
    this.searchQuery = null
    this.mustFilterQuery = null
    this.notQuery = null
    this.sorting = null
  }

  destroy() {
    this.fieldMappingAdapter.destroy()

    this.index = null
    this.type = null
    this.fieldMapping = null
    this.fieldMappingAdapter = null
    this.searchQuery = null
    this.mustFilterQuery = null
    this.notQuery = null
    this.sorting = null
  }

  /*------------------------------METHODS------------------------------*/
  buildKeywords(keywords) {
    if (!this.notQuery) this.notQuery = []
    if (!this.searchQuery) this.searchQuery = []

    let queryGroup
    // build keywords
    keywords.forEach(({ field, operator, keyword, type }) => {
      const mappedField = this.fieldMappingAdapter.getMappedField(field)
      if (operator === Operators.NOT) {
        this.notQuery.push({
          match_phrase: { [mappedField]: keyword }
        })
      } else {
        let queryObject
        if (type === SearchTypes.EXACT) {
          queryObject = {
            match_phrase: { [mappedField]: keyword }
          }
        } else {
          queryObject = {
            match: {
              [mappedField]: {
                query: keyword,
                minimum_should_match: getMinimumByType(type)
              }
            }
          }
        }

        // init querygroup
        if (operator === Operators.AND || !queryGroup) {
          queryGroup = {
            bool: {
              should: []
            }
          }
          this.searchQuery.push(queryGroup)
        }

        queryGroup.bool.should.push(queryObject)
      }
    })
  }

  buildInclude(include) {
    const includeReturn = {}
    include.forEach(({ field, keyword }) => {
      const includedField = this.fieldMappingAdapter.getMappedField(field, true)
      if (!includeReturn[includedField]) includeReturn[includedField] = []
      includeReturn[includedField].push({
        match_phrase: { [includedField]: keyword }
      })
    })

    return Object.keys(includeReturn).map(field => ({
      bool: {
        should: includeReturn[field]
      }
    }))
  }

  buildExclude(exclude) {
    return exclude.map(({ field, keyword }) => ({
      match_phrase: {
        [this.fieldMappingAdapter.getMappedField(field)]: keyword
      }
    }))
  }

  buildMisc(misc) {
    if (!this.notQuery) this.notQuery = []
    if (!this.mustFilterQuery) this.mustFilterQuery = []
    if (!this.searchQuery) this.searchQuery = []

    misc &&
      misc.forEach(elem => {
        const miscField = this.fieldMappingAdapter.getMiscField(elem)

        if (
          elem === MiscTypes.BLANK_CONSIGNEES_EXCLUDED ||
          elem === MiscTypes.BLANK_SHIPPERS_EXCLUDED
        ) {
          this.searchQuery.push({
            exists: {
              field: miscField
            }
          })
        } else if (elem === MiscTypes.MASTER_SHIPMENTS_EXCLUDED) {
          this.notQuery.push({
            match_phrase: { [miscField]: 'Master' }
          })
        } else if (elem === MiscTypes.HOUSE_SHIPMENTS_ONLY) {
          this.mustFilterQuery.push({
            bool: {
              should: [
                {
                  match_phrase: { [miscField]: 'H' }
                },
                {
                  match_phrase: { [miscField]: 'House' }
                }
              ]
            }
          })
        } else if (elem === MiscTypes.MASTER_SHIPMENTS_ONLY) {
          this.mustFilterQuery.push({
            match_phrase: { [miscField]: 'Master' }
          })
        }
      })

    return {
      notQuery: this.notQuery,
      mustFilterQuery: this.mustFilterQuery,
      searchQuery: this.searchQuery
    }
  }

  buildSorting(sortBy) {
    let sortField = 'Arrival Date'
    let sortOrder = 'asc'
    this.sorting = []

    if (sortBy && sortBy.length > 0) {
      if (sortBy.charAt(0) === '-') {
        sortOrder = 'desc'
        sortField = this.fieldMappingAdapter.getMappedField(
          sortBy.substr(1),
          true
        )
      } else if (sortBy.charAt(0) === '+') {
        sortOrder = 'asc'
        sortField = this.fieldMappingAdapter.getMappedField(
          sortBy.substr(1),
          true
        )
      } else {
        sortOrder = 'asc'
        sortField = this.fieldMappingAdapter.getMappedField(sortBy, true)
      }
    }

    this.sorting.push(
      {
        [sortField]: {
          order: sortOrder
        }
      },
      '_score'
    )

    return this
  }

  buildQuery(
    keywords,
    include,
    exclude,
    misc,
    sortBy,
    limit,
    page,
    arrivalDates = ''
  ) {
    let [gte = '', lte = ''] = (arrivalDates && arrivalDates.split('-')) || []
    this.searchQuery = []
    this.mustFilterQuery = []
    this.notQuery = []

    this.buildKeywords(keywords)

    if (include) this.mustFilterQuery.push(...this.buildInclude(include))
    if (exclude) this.notQuery.push(...this.buildExclude(exclude))

    this.buildMisc(misc)
    this.buildSorting(sortBy)

    const querySearch = {
      size: limit,
      from: page * limit - limit,
      track_scores: true,
      sort: this.sorting,
      query: {
        bool: {
          must: this.searchQuery,
          must_not: this.notQuery,
          filter: [
            ...this.mustFilterQuery,
            {
              range: {
                'Arrival Date': {
                  gte: gte,
                  lte: lte
                }
              }
            }
          ]
        }
      }
    }

    return {
      index: this.index,
      type: this.type,
      body: querySearch
    }
  }
}

export default SearchBuilder
