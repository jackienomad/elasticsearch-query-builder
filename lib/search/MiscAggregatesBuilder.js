import { FieldMappingAdapter } from './FieldMappingAdapter'
import { MiscTypes } from './searchUtils'

export class MiscAggregatesBuilder {
  static generateFilterObject(misc) {
    let filter
    if (misc === MiscTypes.BLANK_CONSIGNEES_EXCLUDED) {
      filter = {
        exists: {
          field: 'Consignee'
        }
      }
    } else if (misc === MiscTypes.BLANK_SHIPPERS_EXCLUDED) {
      filter = {
        exists: {
          field: 'Shipper'
        }
      }
    } else if (misc === MiscTypes.MASTER_SHIPMENTS_EXCLUDED) {
      return {
        bool: {
          must_not: {
            match_phrase: {
              Consignee: 'Master'
            }
          }
        }
      }
    } else if (misc === MiscTypes.HOUSE_SHIPMENTS_ONLY) {
      filter = {
        bool: {
          should: [
            {
              match_phrase: {
                'Bill Type': 'H'
              }
            },
            {
              match_phrase: {
                'Bill Type': 'House'
              }
            }
          ]
        }
      }
    } else if (misc === MiscTypes.MASTER_SHIPMENTS_ONLY) {
      filter = {
        bool: {
          must: {
            match_phrase: {
              'Bill Type': 'Master'
            }
          }
        }
      }
    }

    return filter
  }

  static build(fields, fieldMapping) {
    const fieldAdapter =
      fieldMapping instanceof FieldMappingAdapter
        ? fieldMapping
        : new FieldMappingAdapter(fieldMapping)

    const aggregations = {}
    fields.forEach(field => {
      aggregations[field] = {
        filter: MiscAggregatesBuilder.generateFilterObject(field)
      }
    })

    return aggregations
  }
}

export default MiscAggregatesBuilder
