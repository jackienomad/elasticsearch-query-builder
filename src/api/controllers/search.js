import { client } from '../models/queryBuilderModel'
import { SearchBuilder } from 'qb/@/search/SearchBuilder'
import { SearchFormatter } from 'qb/@/search/SearchFormatter'

export async function fetchShipments({
  keywords,
  include,
  exclude,
  misc,
  sortBy,
  limit,
  page,
  index,
  type,
  arrivalDate = ''
}) {
  const builder = new SearchBuilder(index, type)
  const query = builder.buildQuery(
    keywords,
    include,
    exclude,
    misc,
    sortBy,
    limit,
    page,
    arrivalDate
  )

  const resp = await client.search(query)
  return {
    total: resp.hits.total,
    max_score: resp.hits.max_score,
    data: {
      items: SearchFormatter.formatResponse(resp)
    }
  }
}

export default fetchShipments
