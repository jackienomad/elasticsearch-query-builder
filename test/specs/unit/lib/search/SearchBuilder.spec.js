import { SearchBuilder } from 'qb/@/search/SearchBuilder'
import { MiscTypes } from 'qb/@/search/searchUtils'
import fieldMapping from './dataset/sampleFieldMapping'

describe('SearchBuilder class', () => {
  describe('constructor and destructor', () => {
    it('should be instantiable', () => {
      expect(() => {
        new SearchBuilder('sample index', 'sample type', fieldMapping)
      }).not.toThrowError()
    })

    it('should be destroyable', () => {
      const builder = new SearchBuilder(
        'sample index',
        'sample type',
        fieldMapping
      )
      expect(() => {
        builder.destroy()
      }).not.toThrowError()
    })
  })

  describe('buildKeywords', () => {
    const builder = new SearchBuilder(
      'sample index',
      'sample type',
      fieldMapping
    )

    afterAll(() => {
      builder.destroy()
      builder = null
    })

    it('should return mapped include object', () => {
      const actual = builder.buildKeywords([
        {
          field: 'Consignee',
          keyword: 'Chicago',
          operator: 'and',
          type: 'fuzzy'
        },
        {
          field: 'Consignee',
          keyword: 'Arizona',
          operator: 'or',
          type: 'regular'
        },
        {
          field: 'Shipper',
          keyword: 'Chicago',
          operator: 'and',
          type: 'exact'
        },
        ,
        {
          field: 'ProductDescription',
          keyword: 'apple',
          operator: 'not',
          type: 'regular'
        },
        {
          field: 'ProductDescription',
          keyword: 'apple juice',
          operator: 'or',
          type: 'exact'
        }
      ])

      expect(builder.searchQuery).toMatchSnapshot()
      expect(builder.notQuery).toMatchSnapshot()
    })
  })

  describe('buildInclude', () => {
    const builder = new SearchBuilder(
      'sample index',
      'sample type',
      fieldMapping
    )

    afterAll(() => {
      builder.destroy()
      builder = null
    })

    it('should return mapped include object', () => {
      const actual = builder.buildInclude([
        {
          field: 'ConsigneeCity',
          keyword: 'Chicago'
        },
        {
          field: 'ConsigneeState',
          keyword: 'Arizona'
        },
        {
          field: 'ConsigneeState',
          keyword: 'Texas'
        },
        {
          field: 'ConsigneeZipCode',
          keyword: '90210'
        },
        {
          field: 'ConsigneeZipCode',
          keyword: '88410'
        },
        {
          field: 'ShipperAddress',
          keyword: 'Genius Drive'
        },
        {
          field: 'CountryOfOrigin',
          keyword: 'ph'
        },
        {
          field: 'DestinationPort',
          keyword: 'Port of Los Angeles'
        },
        {
          field: 'ForeignPort',
          keyword: 'Port of Busan'
        }
      ])

      expect(actual).toMatchSnapshot()
    })
  })

  describe('buildExclude', () => {
    const builder = new SearchBuilder(
      'sample index',
      'sample type',
      fieldMapping
    )

    afterAll(() => {
      builder.destroy()
      builder = null
    })

    it('should return mapped exclude object', () => {
      const actual = builder.buildExclude([
        {
          field: 'ConsigneeCity',
          keyword: 'Chicago'
        },
        {
          field: 'ConsigneeState',
          keyword: 'Arizona'
        },
        {
          field: 'ConsigneeZipCode',
          keyword: '90210'
        },
        {
          field: 'ShipperAddress',
          keyword: 'Genius Drive'
        },
        {
          field: 'CountryOfOrigin',
          keyword: 'ph'
        },
        {
          field: 'DestinationPort',
          keyword: 'Port of Los Angeles'
        },
        {
          field: 'ForeignPort',
          keyword: 'Port of Busan'
        }
      ])

      expect(actual).toMatchSnapshot()
    })
  })

  describe('buildMisc', () => {
    const builder = new SearchBuilder(
      'sample index',
      'sample type',
      fieldMapping
    )

    afterAll(() => {
      builder.destroy()
      builder = null
    })

    it('should return mapped misc objects', () => {
      const actual = builder.buildMisc([
        MiscTypes.BLANK_CONSIGNEES_EXCLUDED,
        MiscTypes.MASTER_SHIPMENTS_EXCLUDED,
        MiscTypes.HOUSE_SHIPMENTS_ONLY,
        MiscTypes.MASTER_SHIPMENTS_ONLY
      ])

      expect(builder.mustFilterQuery).toMatchSnapshot()
      expect(builder.notQuery).toMatchSnapshot()
    })
  })

  describe('buildSorting', () => {
    const builder = new SearchBuilder(
      'sample index',
      'sample type',
      fieldMapping
    )

    afterAll(() => {
      builder.destroy()
      builder = null
    })

    it('should return default sort object if empty', () => {
      const actual = builder.buildSorting().sorting

      expect(actual).toMatchSnapshot()
    })

    it('should handle descending sort', () => {
      const actual = builder.buildSorting('-Consignee').sorting

      expect(actual).toMatchSnapshot()
    })

    it('should handle ascending sort', () => {
      const actual = builder.buildSorting('+Shipper').sorting

      expect(actual).toMatchSnapshot()
    })

    it('should handle no specified order', () => {
      const actual = builder.buildSorting('MarksAndNumbers').sorting

      expect(actual).toMatchSnapshot()
    })
  })

  describe('buildQuery', () => {
    const builder = new SearchBuilder(
      'sample index',
      'sample type',
      fieldMapping
    )

    afterAll(() => {
      builder.destroy()
      builder = null
    })

    it('should build an es query', () => {
      const actual = builder.buildQuery(
        [
          {
            field: 'Consignee',
            keyword: 'Chicago',
            operator: 'and',
            type: 'fuzzy'
          },
          {
            field: 'Shipper',
            keyword: 'Chicago',
            operator: 'and',
            type: 'exact'
          },
          {
            field: 'Shipper',
            keyword: 'Chicago',
            operator: 'or',
            type: 'regular'
          },
          {
            field: 'ProductDescription',
            keyword: 'orange',
            operator: 'and',
            type: 'exact'
          },
          {
            field: 'ProductDescription',
            keyword: 'apple',
            operator: 'not',
            type: 'exact'
          }
        ],
        [
          {
            field: 'ConsigneeCity',
            keyword: 'Chicago'
          },
          {
            field: 'ConsigneeCity',
            keyword: 'Los Angeles'
          },
          {
            field: 'ConsigneeState',
            keyword: 'Arizona'
          }
        ],
        [
          {
            field: 'CountryOfOrigin',
            keyword: 'ph'
          },
          {
            field: 'DestinationPort',
            keyword: 'Port of Los Angeles'
          }
        ],
        [
          MiscTypes.BLANK_CONSIGNEES_EXCLUDED,
          MiscTypes.MASTER_SHIPMENTS_EXCLUDED,
          MiscTypes.HOUSE_SHIPMENTS_ONLY,
          MiscTypes.MASTER_SHIPMENTS_ONLY
        ],
        '-Shipper',
        20,
        1,
        '1512938867-1544475436'
      )

      expect(actual).toMatchSnapshot()
    })
  })
})
