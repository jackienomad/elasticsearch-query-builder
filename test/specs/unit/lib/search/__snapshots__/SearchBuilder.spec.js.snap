// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBuilder class buildExclude should return mapped exclude object 1`] = `
Array [
  Object {
    "match_phrase": Object {
      "Consignee.Cleaned Address Breakdown.PlaceName": "Chicago",
    },
  },
  Object {
    "match_phrase": Object {
      "Consignee.Cleaned Address Breakdown.StateName": "Arizona",
    },
  },
  Object {
    "match_phrase": Object {
      "Consignee.Cleaned Address Breakdown.ZipCode": "90210",
    },
  },
  Object {
    "match_phrase": Object {
      "Shipper.Address": "Genius Drive",
    },
  },
  Object {
    "match_phrase": Object {
      "Country of Origin": "ph",
    },
  },
  Object {
    "match_phrase": Object {
      "Port of Unlading": "Port of Los Angeles",
    },
  },
  Object {
    "match_phrase": Object {
      "Foreign Port of Lading": "Port of Busan",
    },
  },
]
`;

exports[`SearchBuilder class buildInclude should return mapped include object 1`] = `
Array [
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Consignee.Cleaned Address Breakdown.PlaceName.untouched": "Chicago",
          },
        },
      ],
    },
  },
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Consignee.Cleaned Address Breakdown.StateName.untouched": "Arizona",
          },
        },
        Object {
          "match_phrase": Object {
            "Consignee.Cleaned Address Breakdown.StateName.untouched": "Texas",
          },
        },
      ],
    },
  },
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Consignee.Cleaned Address Breakdown.ZipCode.untouched": "90210",
          },
        },
        Object {
          "match_phrase": Object {
            "Consignee.Cleaned Address Breakdown.ZipCode.untouched": "88410",
          },
        },
      ],
    },
  },
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Shipper.Address.untouched": "Genius Drive",
          },
        },
      ],
    },
  },
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Country of Origin.untouched": "ph",
          },
        },
      ],
    },
  },
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Port of Unlading.untouched": "Port of Los Angeles",
          },
        },
      ],
    },
  },
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Foreign Port of Lading.untouched": "Port of Busan",
          },
        },
      ],
    },
  },
]
`;

exports[`SearchBuilder class buildKeywords should return mapped include object 1`] = `
Array [
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match": Object {
            "Consignee.Name": Object {
              "minimum_should_match": "75%",
              "query": "Chicago",
            },
          },
        },
        Object {
          "match": Object {
            "Consignee.Name": Object {
              "minimum_should_match": "100%",
              "query": "Arizona",
            },
          },
        },
      ],
    },
  },
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Shipper.Name": "Chicago",
          },
        },
        Object {
          "match_phrase": Object {
            "Containers.Cargoes.Description": "apple juice",
          },
        },
      ],
    },
  },
]
`;

exports[`SearchBuilder class buildKeywords should return mapped include object 2`] = `
Array [
  Object {
    "match_phrase": Object {
      "Containers.Cargoes.Description": "apple",
    },
  },
]
`;

exports[`SearchBuilder class buildMisc should return mapped misc objects 1`] = `
Array [
  Object {
    "bool": Object {
      "should": Array [
        Object {
          "match_phrase": Object {
            "Bill Type": "H",
          },
        },
        Object {
          "match_phrase": Object {
            "Bill Type": "House",
          },
        },
      ],
    },
  },
  Object {
    "match_phrase": Object {
      "Bill Type": "Master",
    },
  },
]
`;

exports[`SearchBuilder class buildMisc should return mapped misc objects 2`] = `
Array [
  Object {
    "match_phrase": Object {
      "Bill Type": "Master",
    },
  },
]
`;

exports[`SearchBuilder class buildQuery should build an es query 1`] = `
Object {
  "body": Object {
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match_phrase": Object {
                    "Consignee.Cleaned Address Breakdown.PlaceName.untouched": "Chicago",
                  },
                },
                Object {
                  "match_phrase": Object {
                    "Consignee.Cleaned Address Breakdown.PlaceName.untouched": "Los Angeles",
                  },
                },
              ],
            },
          },
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match_phrase": Object {
                    "Consignee.Cleaned Address Breakdown.StateName.untouched": "Arizona",
                  },
                },
              ],
            },
          },
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match_phrase": Object {
                    "Bill Type": "H",
                  },
                },
                Object {
                  "match_phrase": Object {
                    "Bill Type": "House",
                  },
                },
              ],
            },
          },
          Object {
            "match_phrase": Object {
              "Bill Type": "Master",
            },
          },
          Object {
            "range": Object {
              "Arrival Date": Object {
                "gte": "1512938867",
                "lte": "1544475436",
              },
            },
          },
        ],
        "must": Array [
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match": Object {
                    "Consignee.Name": Object {
                      "minimum_should_match": "75%",
                      "query": "Chicago",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match_phrase": Object {
                    "Shipper.Name": "Chicago",
                  },
                },
                Object {
                  "match": Object {
                    "Shipper.Name": Object {
                      "minimum_should_match": "100%",
                      "query": "Chicago",
                    },
                  },
                },
              ],
            },
          },
          Object {
            "bool": Object {
              "should": Array [
                Object {
                  "match_phrase": Object {
                    "Containers.Cargoes.Description": "orange",
                  },
                },
              ],
            },
          },
          Object {
            "exists": Object {
              "field": "Consignee.Name",
            },
          },
        ],
        "must_not": Array [
          Object {
            "match_phrase": Object {
              "Containers.Cargoes.Description": "apple",
            },
          },
          Object {
            "match_phrase": Object {
              "Country of Origin": "ph",
            },
          },
          Object {
            "match_phrase": Object {
              "Port of Unlading": "Port of Los Angeles",
            },
          },
          Object {
            "match_phrase": Object {
              "Bill Type": "Master",
            },
          },
        ],
      },
    },
    "size": 20,
    "sort": Array [
      Object {
        "Shipper.Name.untouched": Object {
          "order": "desc",
        },
      },
      "_score",
    ],
    "track_scores": true,
  },
  "index": "sample index",
  "type": "sample type",
}
`;

exports[`SearchBuilder class buildSorting should handle ascending sort 1`] = `
Array [
  Object {
    "Shipper.Name.untouched": Object {
      "order": "asc",
    },
  },
  "_score",
]
`;

exports[`SearchBuilder class buildSorting should handle descending sort 1`] = `
Array [
  Object {
    "Consignee.Name.untouched": Object {
      "order": "desc",
    },
  },
  "_score",
]
`;

exports[`SearchBuilder class buildSorting should handle no specified order 1`] = `
Array [
  Object {
    "Containers.Marks.untouched": Object {
      "order": "asc",
    },
  },
  "_score",
]
`;

exports[`SearchBuilder class buildSorting should return default sort object if empty 1`] = `
Array [
  Object {
    "Arrival Date": Object {
      "order": "asc",
    },
  },
  "_score",
]
`;
